### PR TITLE
[Feature] Support UniFormerV1&V2

### PR DESCRIPTION
**Pull request by Andy1621**
## Motivation

- [UniFormerV1](https://github.com/Sense-X/UniFormer/tree/main/video_classification) is an efficient video transformer. 
- [UniFormerV2](https://github.com/OpenGVLab/UniFormerV2) proposes efficient modules for arming vanilla ViT as powerful spatiotemporal learners, and it achieves SOTA performances on various benchmarks.

## Modification

- `uniformer.py` and `uniformerv2.py` in `mmaction/models/backbones`.
- Add `dropout` in `mmaction/models/heads/timesformer_head.py`.
- Add multi-clip testing for uniform sampling in `mmaction/datasets/transforms/loading.py`.
- Support different delimiters (like ',') in `mmaction/datasets/video_dataset.py`.
- Add testing configs for UniFormerV1&V2.

## BC-breaking (Optional)

The modification introduces does not change that break the back-compatibility of this repo.

### CODE REVIEW

File: mmaction/datasets/video_dataset.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ class VideoDataset(BaseActionDataset):
             Defaults to ``RGB``.
         test_mode (bool): Store True when building test or validation dataset.
             Defaults to False.
+        delimiter (str):
```
Comment (cir7) [inline mmaction/datasets/video_dataset.py:51]: please add argument docstring

Comment (Andy1621) [reply mmaction/datasets/video_dataset.py]: Done

File: mmaction/models/backbones/uniformer.py [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+        B, _, T, H, W = x.shape
+        x = x.flatten(2).transpose(1, 2)
+        x = self.norm(x)
+        x = x.reshape(B, T, H, W, -1).permute(0, 4, 1, 2, 3).contiguous()
+        return x
+
+
+@MODELS.register_module()
+class UniFormer(BaseModule):
+    """UniFormer.
```
Comment (cir7) [inline mmaction/models/backbones/uniformer.py:473]: please add paper link for Uniformer, take [swin](https://github.com/open-mmlab/mmaction2/blob/dev-1.x/mmaction/models/backbones/swin.py#L707) as a reference.

File: mmaction/models/backbones/uniformerv2.py [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+
+        weight = self.sigmoid(self.balance)
+        residual = x.view(L, N, T_down, C)[0].mean(1)  # L, N, T, C
+        out = self.norm((1 - weight) * cls_token[0, :, :] + weight * residual)
+        return out
+
+
+@MODELS.register_module()
+class UniFormerV2(BaseModule):
+    """UniFormerV2:
```
Comment (cir7) [inline mmaction/models/backbones/uniformerv2.py:395]: please add paper link

File: mmaction/models/heads/timesformer_head.py [hunk 4]
```diff
@@ -17,6 +17,7 @@ class TimeSformerHead(BaseHead):
         loss_cls (dict or ConfigDict): Config for building loss.
             Defaults to `dict(type='CrossEntropyLoss')`.
         init_std (float): Std value for Initiation. Defaults to 0.02.
+        dropout_ratio (float): Probability of dropout layer. Default: 0.0.
```
Comment (cir7) [inline mmaction/models/heads/timesformer_head.py:20]: Default: 0.0 => Defaults to 0.0

File: mmaction/models/backbones/uniformer.py [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+            weight_3d = torch.zeros(*weight_2d.shape)
+            weight_3d = weight_3d.unsqueeze(2).repeat(1, 1, time_dim, 1, 1)
+            middle_idx = time_dim // 2
+            weight_3d[:, :, middle_idx, :, :] = weight_2d
+        else:
+            weight_3d = weight_2d.unsqueeze(2).repeat(1, 1, time_dim, 1, 1)
+            weight_3d = weight_3d / time_dim
+        return weight_3d
+
+    def _init_weights(self, m):
```
Comment (cir7) [inline mmaction/models/backbones/uniformer.py:607]: As a convention of OpenMMLab repo, we use `init_cfg` to specify weight initialization. For complicated situation `init_cfg` can't cover，we override `init_weights` and automatically call it in `Runner`.

File: mmaction/models/backbones/uniformerv2.py (hunk 3 above)
Comment (Andy1621) [reply mmaction/models/backbones/uniformerv2.py:395]: Done

File: mmaction/models/heads/timesformer_head.py (hunk 4 above)
Comment (Andy1621) [reply mmaction/models/heads/timesformer_head.py]: Done

File: mmaction/models/backbones/uniformer.py (hunk 2 above)
Comment (Andy1621) [reply mmaction/models/backbones/uniformer.py:473]: Done

File: mmaction/models/backbones/uniformer.py (hunk 5 above)
Comment (Andy1621) [reply mmaction/models/backbones/uniformer.py]: Done

File: mmaction/datasets/video_dataset.py [hunk 6]
```diff
@@ -48,6 +48,8 @@ class VideoDataset(BaseActionDataset):
             Defaults to ``RGB``.
         test_mode (bool): Store True when building test or validation dataset.
             Defaults to False.
+        delimiter (str): Delimiter for the annotation file.
+            Defaults to ' ' (whitespace).
```
Comment (ly015) [inline mmaction/datasets/video_dataset.py:52]:
```suggestion
            Defaults to ``' '`` (whitespace).
```

**Review by ly015 — APPROVED** [review]
